oxidation/libparsec/crates/client_connection/src/authenticated_cmds.rs

```diff
@@ -0,0 +1,313 @@
+pub const PARSEC_CONTENT_TYPE: &str = "application/msgpack";
+
+/// Factory that send commands in a authenticated context.
+pub struct AuthenticatedCmds {
+    /// HTTP Client that contain the basic configuration to communicate with the server.
+    client: Client,
+    url: Url,
+    device_id: String,
+    signing_key: SigningKey,
+}
+
+impl AuthenticatedCmds {
+    /// Create a new `AuthenticatedCmds`
+    pub fn new(
+        client: Client,
+        server_url: BackendOrganizationAddr,
+        device_id: DeviceID,
+        signing_key: SigningKey,
+    ) -> Result<Self, url::ParseError> {
+        let url = server_url.to_authenticated_http_url();
+
+        let device_id = base64::encode(device_id.to_string().as_bytes());
+
+        Ok(Self {
+            client,
+            url,
+            device_id,
+            signing_key,
+        })
+    }
+}
+
+macro_rules! impl_auth_cmds {
+    (
+        $(
+            $(#[$outer:meta])*
+            $name:ident($($key:ident: $type:ty),*)
+        )+
+    ) => {
+        $(
+            $(#[$outer])*
+            pub async fn $name(&self, $($key: $type),*) -> command_error::Result<authenticated_cmds::$name::Rep> {
+                let request_builder = self.client.post(self.url.clone());
+
+                let data = authenticated_cmds::$name::Req::new($($key),*).dump()
+                    .expect(concat!("failed to serialize the command ", stringify!($name)));
+
+                let req = prepare_request(request_builder, &self.signing_key, &self.device_id, data).send();
+                let resp = dbg!(req.await)?;
+                if resp.status() != reqwest::StatusCode::OK {
+                    return Err(CommandError::InvalidResponseStatus(resp.status(), resp));
+                }
+
+                let response_body = resp.bytes().await?;
+
+                authenticated_cmds::$name::Rep::load(&response_body).map_err(CommandError::InvalidResponseContent)
+            }
+        )+
+    };
+}
+
+/// Prepare a new request, the body will be added to the Request using [RequestBuilder::body]
+fn prepare_request(
+    request_builder: RequestBuilder,
+    signing_key: &SigningKey,
+    device_id: &str,
+    body: Vec<u8>,
+) -> RequestBuilder {
+    let request_builder = sign_request(request_builder, signing_key, device_id, &body);
+
+    let mut content_headers = HeaderMap::with_capacity(2);
+    content_headers.insert(CONTENT_TYPE, HeaderValue::from_static(PARSEC_CONTENT_TYPE));
+    content_headers.insert(
+        CONTENT_LENGTH,
+        HeaderValue::from_str(&body.len().to_string()).expect("numeric value are valid char"),
+    );
+
+    request_builder.headers(content_headers).body(body)
+}
+
+/// Sing a request by adding specific headers.
+fn sign_request(
+    request_builder: RequestBuilder,
+    signing_key: &SigningKey,
+    device_id: &str,
+    body: &[u8],
+) -> RequestBuilder {
+    let timestamp = chrono::Utc::now().to_string();
```

Comment:
use explicit [to_rfc2822](https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.to_rfc2822) instead of implicit to_string conversion:
- it's much more readable
- to_string uses the blanked implementation that relies on `impl Display for DateTime` which has no guarantee not to change it display format in the future !